rfcs/text/0001_lifecycle_init.md

```diff
@@ -0,0 +1,141 @@
+completely adhere to the order of execution in a lifecycle, plugins have no way
+of saying "execute this only when all plugins have initialized". It's not
+practical for a plugin that has side effects like pushing cluster privileges to
+Elasticsearch to constantly be executing those side effects whenever an
+observable changes. Instead, they need a point in time when they can safely
+assume the necessary configurations have been made.
+
+An `init` lifecycle handler would allow core and plugins to expose contracts
+that have a reliable expiration in the context of the overall lifecycle.
+
+# Detailed design
+
+A new `init` lifecycle handler will be adopted for services and plugins. The
+order in which lifecycle handlers execute will be:
+
+1. `init`
+2. `start`
+3. `stop`
+
+## Core
+
+The core system will have an `init` function that will get executed prior to
+`start`. An `init` function will also be added to all core services, and will
+be invoked from the core `init` in the same spirit of `start` and `stop`.
+
+Decisions on which service functionality should belong in `init` vs `start`
+will need to be handled case-by-case and is beyond the scope of this RFC, but
+much of the existing functionality will likely be exposed through `init`
+instead.
+
+## Plugins
+
+Plugins will have an `init` function that will get executed by the core plugin
+service from its own `init`.
+
+Like `start` and `stop`, the `init` lifecycle handler will receive
+init-specific core contracts via the first argument.
+
+Also like `start` and `stop`, the `init` lifecycle handler will receive the
+init-specific plugin contracts from all plugins that it has a declared
+dependency on via the second argument.
+
+# Drawbacks
+
+- An additional lifecycle handler adds complexity for many plugins and services
+which draw no direct benefit from it.
+- The answer to "does this belong in `init` or `start`?" is not always clear.
+There is not a formal decision tree we can apply to all circumstances.
+- While lifecycle hooks are relatively new, there still many services that will
+need to be updated.
+- Adopting new lifecycle hooks is a slippery slope, and the more we have in the
+system, the more complicated it is to reason about the capabilities of the
+system at any given point.
+
+# Alternatives
+
+When a service or plugin needs to know when initialization has finished, it can
+expose a custom event or transaction system via its relevant contracts so it
+can tell when downstream code has finished initializing. One significant
+drawback to this approach is that it only works when the plugin that needs to
+wait for initialization isn't dependent on an upstream service that does not
+implement a similar transaction capability.
+
+# Adoption strategy
+
+Adoption will need to be manual. Since the bulk of the `start` logic in the
+repo today is configuration-oriented, I recommend renaming `start`->`init` in
+all services and plugins, and then adding an empty `start` where it is
```

Comment:
What do you think about making each lifecycle method optional? That way, plugins that don't need separate init and starts don't have to have that complexity in their exported plugin. Establishing this pattern now would also make adoption easier for future lifecycle additions.

Reply:
I think that is a reasonable discussion to have, but I think that's a separate discussion from this proposal.